sentry_streams/sentry_streams/pipeline/pipeline.py

```diff
@@ -431,3 +463,77 @@ class FlatMap(TransformStep[Any]):
     """
 
     step_type: StepType = StepType.FLAT_MAP
+
+
+######################
+# Complex Primitives #
+######################
+
+
+@dataclass
+class Parser(ComplexStep, WithInput, Generic[TransformFuncReturnType]):
+    """
+    A step to decode bytes, deserialize the resulting message and validate it against the schema
+    which corresponds to the message type provided. The message type should be one which
+    is supported by sentry-kafka-schemas. See examples/ for usage, this step can be plugged in
+    flexibly into a pipeline. Keep in mind, data up until this step will simply be bytes.
+
+    Supports both JSON and protobuf.
+    """
+
+    msg_type: Type[TransformFuncReturnType]
+
+    def convert(self) -> Step:
+        return Map(
+            name=self.name,
+            function=msg_parser,
+        )
+
+
+@dataclass
+class BatchParser(
+    ComplexStep,
+    WithInput,
+    Generic[TransformFuncReturnType],
+):
+    msg_type: Type[TransformFuncReturnType]
+
+    def convert(self) -> Step:
+        return Map(
+            name=self.name,
+            function=batch_msg_parser,
+        )
+
+
+@dataclass
+class Serializer(ComplexStep, WithInput):
+    """
+    A step to serialize and encode messages into bytes. These bytes can be written
+    to sink data to a Kafka topic, for example. This step will need to precede a
+    sink step which writes to Kafka.
+    """
+
+    dt_format: Optional[str] = None
+
+    def convert(self) -> Step:
+        return Map(
+            name=self.name,
+            function=msg_serializer,
+        )
+
+
+@dataclass
+class Reducer(ComplexStep, WithInput, Generic[MeasurementUnit, InputType, OutputType]):
+    window: Window[MeasurementUnit]
+    aggregate_func: Callable[[], Accumulator[Message[InputType], OutputType]]
+    aggregate_backend: AggregationBackend[OutputType] | None = None
+    group_by_key: GroupBy | None = None
+
+    def convert(self) -> Step:
+        return Aggregate(
```

Comment:
I think it is the other way around. `Reduce` is the basic step and `Aggregate` and `Batch` are the complex ones.

Reply:
This is ported verbatim from the `chains.py` API. `Reducer` is semantic sugar for `Aggregate` and this also `Reduce`.